Refactor XarrayZarrRecipe to be serialization-friendly

TomAugspurger opened the pull request:
This is (another) refactor to the XarrayZarrRecipe to resolve some of the memory issues we've seen when executing large jobs (https://github.com/pangeo-forge/pangeo-forge-azure-bakery/issues/10, https://github.com/pangeo-forge/pangeo-forge-recipes/issues/151)

It builds on https://github.com/pangeo-forge/pangeo-forge-recipes/pull/153, which adds `to_dask` and `to_prefect` methods (probably should do that at the base recipe level). 

It looks like a large diff, but it's primarily just moving code from methods on `XarrayZarrRecipe` to top-level functions, and forwarding arguments appropriately. This eliminates `self` from the functions sent to workers. More explanation at https://github.com/pangeo-forge/pangeo-forge-recipes/pull/160/files#diff-e12c886cc124886c5cfa5313d760a36c39649af9da845077c663e6feab8487b5R685-R693. 

The main outstanding task right now is ensuring that the metadata_cache is being handled properly. I need to better understand where that object is supposed to live (on the client or workers?) and who is supposed to be able to write to it (do workers write to it? Is it expected to be global so that writes from a worker process are seen by the client?)

- [x] Move the new compilation methods to `BaseRecipe`. This would also help clarify the API questions (`finalize_target` vs `_finalize_target`, which are required to be implemented where, etc.)
- [x] Remove the `to_pipelines` method
- [x] Implement `to_function` method, which is equivalent to the existing PythonExecutor
- [x] Update executor docs (started in 9c39bf77c332474260d0109ae083f4ffec694df7)
- [x] Rerun all notebooks in `docs/tutorials` to verify no end-user API changes are needed.
- [x] Verify metadata caching isn't broken

Closes https://github.com/pangeo-forge/pangeo-forge-recipes/issues/116.

rabernat commented:
Thanks for working on this Tom!

> It builds on #153,

Does it directly build on #153? Or is it a new implementation of similar ideas.

This is a pretty big refactoring, so I'd like to understand the motivation more clearly, specifically, why this is needed on top of #153.

Is the basic issue that we cannot use _any_ methods (that contain `self` as an argument) whatsoever without embedding the large Recipe objects in every task? So therefore you need to essentially rewrite everything in functional form?

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -46,6 +48,480 @@ def _chunk_metadata_fname(chunk_key) -> str:
+    xarray_open_kwargs: dict,
+    delete_input_encoding: bool,
+    process_input: Optional[Callable[[xr.Dataset, str], xr.Dataset]],
+    metadata_cache: Optional[MetadataTarget],
```
I feel like these long argument blocks may hurt maintainability. There is so much room for programmer error when passing long lists of arguments through the stack. If we go this route, perhaps we want to enclose some of the arguments in a `Config` object?

TomAugspurger replied:
Agreed! Though I *think* the biggest risk of programmer error is right now / in future major refactors. So it comes down to simplicity of the code (not having to look up what is this "FooConfig" thing?) vs. typing all these out. I'm 51% in favor of writing things out explicitly like this, but am happy to go with a Config / Options style object.

I will say that mypy has been helpful here. It caught a couple issues before running the tests.

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -206,303 +682,178 @@ def _set_target_chunks(self):
+        cache_input_task = task(self._cache_input, name="cache_input")
+        prepare_target_task = task(self._prepare_target, name="prepare_target")
+        store_chunk_task = task(self._store_chunk, name="store_chunk")
+        finalize_target_task = task(self._finalize_target, name="finalize_target")
```
Here you not using `_finalize_target` rather than `finalize_target`, etc.

Is there any point in maintaining the `finalize_target` methods if we are going to remove the existing `to_pipelines` method in `BaseRecipe`?

TomAugspurger replied:
I think having the top-level property `.finalize_target` is still really nice for users who are developing / debugging recipes. We want to keep that, while still allowing Dask and Prefect to access the (partially applied) function itself, so they can wrap it in a Task.

rabernat commented:
Ok, so I think I am on board with this as hopefully the right solution to the serializability problems.

Here is a partial checklist of some things that IMO would need to be done before we merge this:

- [ ] Move the new compilation methods to `BaseRecipe`. This would also help clarify the API questions (`finalize_target` vs `_finalize_target`, which are required to be implemented where, etc.)
- [ ] Remove the `to_pipelines` method
- [ ] Implement `to_function` method, which is equivalent to the existing PythonExecutor
- [ ] Update executor docs (started in 9c39bf77c332474260d0109ae083f4ffec694df7)
- [ ] Rerun all notebooks in `docs/tutorials` to verify no end-user API changes are needed.

Happy to help with some of these via pushing directly to this brach.

Thanks again Tom for taking the time to sort out the core issues.

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -206,303 +682,178 @@ def _set_target_chunks(self):
+            cache_input,
+            cache_inputs=self.cache_inputs,
+            input_cache=self.input_cache,
+            file_pattern=self.file_pattern,
```
I am having a hard time understanding why these partial functions are "small" in a serialization sense, since each function in the graph essentially contains a `functools.partial`-wrapped version of all of the same attributes that are part of the recipe class? In particular, `file_pattern` and `target`s are all fairly complex / large objects themselves, which are now being curried into a single-argument function. How is that any better than having a method attached to a class?

TomAugspurger replied:
Just noting that my initial attempt to answer this question have failed. My hunch is that it's just easier to serialize these functions than it is to serialize the dataclass, but I don't I'd like to have a stronger justification than that. Pointing to the workflow in https://github.com/pangeo-forge/pangeo-forge-azure-bakery/issues/10 is some evidence, but it's pretty indirect.

I'll keep trying to come up with a clear answer.

rabernat replied:
To be clear, having a forensic understanding of this is interesting academically but far less important than actually shipping code that works. 😁  So given limited time, I would focus on the checklist above (rather than digging deeper). 

Let me know if you want help on any aspects here. I have stopped working on #153 in the meantime.

TomAugspurger replied:
I'll take care of the first few items around `to_pipelines()` and will look into the outstanding metadata caching.

TomAugspurger replied:
I updated the docs and notebooks. I had some trouble running the notebook, but I think it was an issue with my local internet connection.

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -206,303 +682,178 @@ def _set_target_chunks(self):
-            yield k
+    def to_prefect(self):
+        """Compile the recipe to a Prefect.Flow object."""
+        from prefect import Flow, task, unmapped
```
Isn't this now implemented twice? You also have this in BaseRecipe, no?

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -206,303 +682,178 @@ def _set_target_chunks(self):
-        return xr.open_zarr(target_mapper)
+        for i, input_key in enumerate(self.iter_inputs()):
+            dsk[(f"cache_input-{token}", i)] = (self._cache_input, input_key)
+        dsk[f"checkpoint_0-{token}"] = (lambda *args: None, list(dsk))
```
Same, this is also implemented in BaseRecipe, no?

rabernat reviewed (commented):
Is it intentional to override the `to_dask` and `to_prefect` methods from `BaseRecipe` in `XarrayZarrRecipe`?

TomAugspurger commented:
> Is it intentional to override the to_dask and to_prefect methods from BaseRecipe in XarrayZarrRecipe?

Fixed in 6d190ff5cc87abe296b82498f474688de4f898f9, to remove the `to_*` methods from XarrayZarrRecipe and move them to the base.

That also removes the `@closure` and underscore-versions of the properties (`_prepare_target`, etc). since they aren't needed anymore.

TomAugspurger commented:
Well, I feel a bit silly. The difficulty in answering https://github.com/pangeo-forge/pangeo-forge-recipes/pull/160/files/e70d52662875f8835cc180cb289fc4e6d4445e4a#diff-e12c886cc124886c5cfa5313d760a36c39649af9da845077c663e6feab8487b5 spurred some more investigation into what was actually taking a lot of size to serialize.

It turns out  that most of the size was in the FilePattern class. Maybe this isn't too surprising, but it did surprise me that the size was actually in the function `filepatterns_from_sequence` at https://github.com/pangeo-forge/pangeo-forge-recipes/blob/562e76ccb77831540d5424e948a6fe4ec440263a/pangeo_forge_recipes/patterns.py#L151-L160, rather than the class itself. And then it was obvious, in https://github.com/pangeo-forge/pangeo-forge-azure-bakery/blob/8c7c40183d3a0eedbaab7dac483287b8abf54c46/flow_test/oisst_recipe.py#L100-L106 we create a list with 14,000 (largeish) strings. I'm still not sure why, but apparently multiple instances of that list were being created when the function was deserialized (maybe since it's referenced via a closure, rather than a top-level function?).

The alternative is to create the `FIlePattern` class "manually": https://github.com/pangeo-forge/pangeo-forge-azure-bakery/blob/c70246d209a529ba921b73ed779bfbb57e9178a6/flow_test/oisst_recipe.py#L96-L110. This just requires a function, and a `range` object. Much smaller.

For some reason in my debugging, I had turned `cache_input` off. Which led me to conclude that this branch was helping. I'll now test the alternative FilePattern against pangeo-forge-recipes master, to see if this is even necessary.

rabernat commented:
> It turns out that most of the size was in the FilePattern class. Maybe this isn't too surprising, but it did surprise me that the size was actually in the function `filepatterns_from_sequence` at

:+1: This is definitely expected from my POV. Perhaps we should deprecate that function.

TomAugspurger commented:
> Perhaps we should deprecate that function.

I think that's worthwhile. It is convenient, but I worry that this will bite us again as we scale users's recipes to large jobs. And I hope that manually constructing the FilePattern isn't too much more difficult.

---

FWIW, when I ran the flow using the fixed FilePattern construction on pangeo-forge-recipes master, my scheduler was OOM Killed. It does seem like this branch is doing something.

rabernat commented:
The big question I have is whether this PR is still significantly better than #153 in terms of serialization. (Assuming the more efficient FilePattern approach is used in both cases.)

TomAugspurger commented:
> The big question I have is whether this PR is still significantly better than #153 in terms of serialization.

It seems like it. I ran a flow that just sleeps for the `cache_input` step with pangeo-forge-recipes master, #153, and this PR. #153 still had memory issues on the scheduler, while this PR didn't.

https://github.com/pangeo-forge/pangeo-forge-azure-bakery/issues/10#issuecomment-867165719 has the results.

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -46,6 +47,480 @@ def _chunk_metadata_fname(chunk_key) -> str:
+    # if cache_metadata:
+    #     # if nitems_per_input is not constant, we need to cache this info
+    #     recipe_meta = {"input_sequence_lens": input_sequence_lens}
+    #     return recipe_meta
```
What's going on here? I don't follow your comment.

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -46,6 +47,480 @@ def _chunk_metadata_fname(chunk_key) -> str:
+        process_input=process_input,
+    ) as ds:
+        input_metadata = ds.to_dict(data=False)
+        # TODO(METADATA): set
+        metadata_cache[_input_metadata_fname(input_key)] = input_metadata
```
Is there some reason metadata requires different treatment here?

rabernat commented on pangeo_forge_recipes/recipes/xarray_zarr.py:
```diff
@@ -46,6 +47,480 @@ def _chunk_metadata_fname(chunk_key) -> str:
+
+    if nitems_per_input:
+        input_sequence_lens = (nitems_per_input,) * file_pattern.dims[concat_dim]  # type: ignore
+    # TODO(Tom): Handle metadata caching here
+    # else:
+    # global_metadata = metadata_cache[_GLOBAL_METADATA_KEY]
+    # input_sequence_lens = global_metadata["input_sequence_lens"]
```
Just flagging that metadata stuff has been commented out.

rabernat reviewed (commented):
I went through this again and noticed that all of the global metadata references have been commented out. Can you explain why global metadata is more difficult with this new functional syntax?

Is there something I can do to help here?

TomAugspurger commented:
Fixing the metadata stuff is my (hopefully) last TODO. I'll dig into it today, but maybe you can answer this easily: Where do the writes and reads to the global metadata store happen, client process or the worker processes?

rabernat commented:
Worker processes.

TomAugspurger commented:
Sounds good. So it's assumed that the metadata cache is globally read / writeable, like a blob storage file system. In that case, I think my last commit fixes things, but the tests were passing without it. It's probably worth adding some kind of check to the tests to verify that.


---

I'm running through the notebooks now, and at least some are failing with `TypeError: 'FSSpecTarget' object does not support item assignment.` I'm guessing they're failing on master too. The targets should be a `MetadataTarget`, rather than an `FsspecTarget`. Maybe related to https://github.com/pangeo-forge/pangeo-forge-recipes/pull/133.

I'll update the notebooks to use MetadataTarget.

TomAugspurger commented:
The notebooks caught an issue: I had removed `XarrayZarrRecipe.open_input` and `XarrayZarrRecipe.open_chunk`. These are useful helpers for debugging a recipe so I restored them and added a basic test to ensure they're present.

The notebooks are now updated to use a `MetadataCache` and are all passing. I think this should be good to go.

rabernat reviewed (approved):
This is fantastic. Thanks so much Tom for all your hard work on this!

Could you just update the [release notes](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/master/docs/development/release_notes.md)? I think at this point our next release is 0.4, so probably need to add a new section.

TomAugspurger commented:
Done in https://github.com/pangeo-forge/pangeo-forge-recipes/pull/160/commits/f57e36adc661f55f7b1e9703857228b79674a651.

rabernat commented:
@TomAugspurger, can you think of a test we could add to guard against regressions related to serialization? Relevant for big PRs like #166 which touch a lot of different pieces...

TomAugspurger commented:
Nothing really comes to mind :/ The typical way to check is to serialize the object and then check the size of the bytestring. But IIRC we found that these objects weren't all that large.
